---
title: "Introduction to human popgen"
author: "Vladimir Bajic"
date: "04.05.2020 - 05.05.2020"
output:
  rmdformats::readthedown:
    self_contained: true
    thumbnails: true
    lightbox: true
    gallery: false
    highlight: tango
    toc_depth: 5
---


```{r setup, echo=FALSE, cache=FALSE}
library(knitr)
library(rmdformats)

## Global options
options(max.print="75")
opts_chunk$set(echo=FALSE,
	             cache=TRUE,
               prompt=FALSE,
               tidy=TRUE,
               comment=NA,
               message=FALSE,
               warning=FALSE,
               engine.path = list(bash = 'C:\\WINDOWS\\SYSTEM32\\bash.exe'))
opts_knit$set(width=75)
```

____________________________________________________________________
<br /><br /><br />

# Studying Human Diversity

Studying human diversity raises important ethical and methodological questions, and we will only mention some of them. Observations and inquiries in the area of human diversity are probably as old as humanity, and have in the past been linked to both blatant and subtle forms of __racism__.

<br /><br /> _Should we study human genetic diversity at all, or is this an area of work where the potential for misuse outweighs the potential benefits to such an extent that it should not be pursued?_ <br />

Such studies already have a long history, and a pragmatic answer to this question is that information on human genetic diversity is needed, and is therefore generated, for __medical__ and __forensic__ applications, and thus is already available whatever evolutionary geneticists decide to do, so we must be ready to consider its implications and consequences. Furthermore, genetic information, in fact, refutes any scientific basis for racism as the existence of discrete human groups. Most human genetic variation is found __within__ any individual population, except for a few loci affected by natural selection. It can therefore be used to argue that racism-the belief that discrimination between apparent groups is justifiable-is an entirely __social construct__. In addition, genetic information is of enormous intrinsic interest to many people, not just scientists.

********************************************
Benefits from genetic diversity studies are:

* increased understanding of genetic history and relationships; 

* medical advances such as the identification of genes predisposing to disease; 

* accurate paternity testing, victim and assailant identification, and other forensic applications; 

* sometimes, immediate benefits to the population, such as medical advice or treatment. However, complications also arise, for example because the people who receive most of the long-term benefits may not be the donors.

********************************************

Outstanding issues that have not been fully resolved include:

* Is informed consent from members of cultures that do not ascribe to Western scientific values truly "informed"? Indeed, can even leading geneticists such as Jim Watson and Craig Venter, who have volunteered to have their whole genomes sequenced and made public, appreciate the full implications when these may only become apparent in the future as research reveals the medical implications of DNA variants?

* How much information about the donor should accompany a cell line or DNA sample, so that the privacy of the donor is not infringed?

* Can samples collected with no written consent many years ago, or perhaps decades ago, still be used?

* Can samples collected for one study be used in another?

* Can an individual give broad consent for all future studies, which may involve techniques that do not yet exist and have implications that are not currently understood (related to the first point, above)? 

It is difficult to give general answers to many of the ethical questions that diversity studies raise; indeed, the possibility of ever more comprehensive genetic studies is one of the driving forces in the field of medical ethics. Answers may emerge more satisfactorily through the consideration of individual cases than through prior reasoning based on principles.

********************************************

<br /><br /> _Who should be studied?_ <br />

Sampling always creates problems: is the sample appropriate and representative? If not, conclusions drawn from the sample may not be applicable to the rest of the population that was sampled. Analyzing everyone would avoid the complications introduced by sampling, and some have argued that it is fairer, but at present it is impractical for DNA studies and even for DNA-free genetics based, for example, on phenotypic traits. This is likely to remain true for the foreseeable future, so the issues raised by sampling must be addressed. Although human genetic diversity has been investigated for a long time, the early studies using pre-DNA polymorphisms aroused little controversy or public interest. Attitudes changed with the launch of the __Human Genome Diversity Project (HGDP)__ in 1991, and all subsequent large-scale projects, including the __HapMap__, __Genographic__, and __1000 Genomes Projects__, have been influenced by this legacy.

*********************************************************************

## Large-scale studies of human genetic variation and publicly available datasets 

A few large-scale studies of human genetic variation have made major contributions to human evolutionary genetics. A number of different factors influence the utility of a given dataset or resource to different areas of study within human genetics. The primary utility of population-genetic datasets to _medical genetics_ studies is providing information on the frequency of particular alleles in populations of broadly-defined ancestry and in constituting a resource for phasing and imputation, such that the sample size of the dataset is key. For _population history_ studies, the diversity of the sampled population is very important, as different populations might provide different information into history. Here we will mention only a few of them, and we will focus on those that provide access to data from many different and diverse human populations as this is clearly key to efforts to understand human genetic diversity and population history. 

*********************************************************************

[HapMap](https://www.ncbi.nlm.nih.gov/probe/docs/projhapmap/) <br />
The HapMap project was an early effort that provided array genotypes from populations of mainly European, East Asian, and African ancestry (International HapMap Consortium 2005). It allows researchers to find genes and genetic variations that affect health and disease. The original mission statement of the International HapMap Project was to develop a haplotype map of the human genome, HapMap, which would describe the common patterns of human DNA sequence variation. Through this research, millions of SNPs were discovered and many GWAS studies used this dataset in research for disease association. This project was a necessary stepping stone for the 1KG project which utilizes many of the same populations.  While this project was an impactful start for the scientific community, the HapMap Project has lost momentum in research (statement from Buchanan et al. 2012). The number of novel variants is constantly increasing and many believe that the 1000 Genomes Project could potentially overshadow the utility of HapMap.

*********************************************************************

[The Human Genome Diversity Project (HGDP)](http://www.cephb.fr/en/hgdp_panel.php)<br />
and the CEPH foundation established a collection of cell lines from a large number of diverse human populations (Cann, et al. 2002), and array genotyping of this collection has resulted in widely used datasets (Jakobsson, et al. 2008; Li, et al. 2008; Patterson, et al. 2012). The HGDP-CEPH collection has several attractive features as a resource for human population genetics. It contains individuals from about 52 different populations, with typically around 20 individuals from each of these. While projects such as HapMap and the 1000 Genomes Project sampled individuals mainly from major continental populations (sampling criteria included being non-vulnerable and relevant to medical-genetic studies), the HGDP sampling encompasses many smaller populations of particular anthropological, linguistic, historical or genetic interest. As a few examples, from Africa, it contains the Khoe-San, believed to represent the earliest branching modern human population, and central African rain forest hunter-gatherers. From Europe, it contains the Basque population, one of the few European groups to speak a language that is not part of the Indo-European family, and the isolated Orkney islanders. From the Middle East and South Asia, it contains the Druze ethno-religious group and the isolated Kalash group from the western Himalayas. From East Asia, it contains several minority ethnic groups from China, including the Turkic speaking Uyghurs from western China. From Oceania, it importantly contains Papua New Guineans. From the Americas, it contains several Native American groups without European admixture, which otherwise is ubiquitous in the majority populations. As the resource consists of cell lines, an unlimited amount of DNA can be obtained. Additionally, while data obtained for population history studies sometimes has restrictions on how it can be used or shared, the policies of the HGDP-CEPH resource is such that all data can be analyzed without restrictions and distributed openly. To date, array genotype and other data generated from this collection have been used in hundreds of population genetics studies.

*********************************************************************

[Population Reference Sample (PORES)](https://www.ncbi.nlm.nih.gov/projects/gap/cgi-bin/study.cgi?study_id=phs000145.v4.p2) <br />
The POPRES is a resource for population, disease, and pharmacological genetics research. It contains genotype array data on close to 6,000 individuals (Nelson, et al. 2008), with particularly good geographical coverage of Europe; however, its wider usage has been limited as the data are available only under managed access rather than open access. The genotype and demographic data from this reference sample are freely available through the NCBI database of Genotypes and Phenotypes (dbGaP).

*********************************************************************

[The 1000 Genomes Project (1KGP)](https://www.internationalgenome.org/about) <br />
The 1KGP released data based on a combination of low-coverage whole-genome and high-coverage exome sequencing data from worldwide human populations, initially from seven populations (1000 Genomes Project Consortium 2010), then 14 populations (1000 Genomes Project Consortium 2012) and finally 26 populations (1000 Genomes Project Consortium 2015).

********************************

The aggregation of data generated in many different studies, particularly large __medical genetics__ studies, has led to the establishment of resources based on data from tens of thousands of individuals.

[The Genome Aggregation Database (gnomAD)](https://gnomad.broadinstitute.org/)  <br />
gnomAD is a coalition of investigators seeking to aggregate and harmonize __exome__ and __genome sequencing data__ from a variety of large-scale sequencing projects, and to make summary data available for the wider scientific community. In its first release, which contained exclusively exome data, it was known as the Exome Aggregation Consortium (ExAC) (see Lek, et al. 2016). 

[The Haplotype Reference Consortium (HRC)](http://www.haplotype-reference-consortium.org/) <br />
While the donor consents for the constituent studies typically preclude open access to the individual genotypes, these resources enable accurate assessments of allele frequencies and/or high-quality genotype phasing and imputation (see McCarthy, et al. 2016). 


********************************

Recently, projects with sampling strategies focused explicitly on __population history__ have provided high-coverage whole-genome sequencing data from very large numbers of diverse populations, but limited to typically 2-4 individuals from each of them.

[The Simons Genome Diversity Project (SGDP)](https://www.simonsfoundation.org/simons-genome-diversity-project/) <br />
Complete genome sequences from more than one hundred diverse human populations. The sampling strategy differs from studies of human genome diversity that have aimed to maximize medical relevance by studying populations with large numbers of present-day people. This new study takes a different approach by sampling populations in a way that represents as much anthropological, linguistic and cultural diversity as possible, and thus includes many deeply divergent human populations that are not well represented in other datasets. All genomes in the dataset were sequenced to at least 30x coverage using Illumina technology.

[The Estonian Biocentre Human Genome Diversity Panel (EGDP)](www.ebc.ee/free_data) <br />
represents a dataset of 483 individuals from 148 populations worldwide, including 379 new genomes from 125 populations, which we group into Diversity and Selection Sets (Pagani, et al. 2016).

********************************


<br /><br /><br />

# Exercise dataset

For this exercise we are going to work with [HGDP-CEPH Dataset 11](http://www.cephb.fr/en/hgdp_panel.php) submitted by Harvard Genetic Department.

This dataset contains SNPs from different human populations across the globe. The aim of this exercise is to get familiar with a few widely used analyses in human popgen as well as to understand  genetic variation in human populations. The dataset we chose is genotyped on the Affymetrix Axiom Human Origins Array which was designed to reduce problems of [ascertainment bias](https://www.ncbi.nlm.nih.gov/books/NBK9792/). Effect of the ascertainment bias has been documented for most of the commercially available SNP array platform. 

_In the figure below each line represents a chromosome, and each dot represents a variant. The usage on population 2 of a set of markers optimized to detect the genetic diversity of population 1 (B) underestimates the higher diversity of population 2 (A)._ [Taken from Luca Pagani Research gate](https://www.researchgate.net/publication/303388742_Through_the_layers_of_the_Ethiopian_genome_a_survey_of_human_genetic_variation_based_on_genome-wide_genotyping_and_re-sequencing_data/figures?lo=1)

<center>
![Ascertainment bias](C:\Users/Vladimir Bajic/Documents/Nowick_Lab/Master_course_Bioinformatics_for_Biologists/SoSe2020/popge_intro/figures_for_html/Effect-of-the-ascertainment-bias-documented-for-most-of-the-commercially-available-SNP_W640.jpg){width=70%}
</center>

*********************************************************************

## Affymetrix Axiom Human Origins Array

Although the __SNP arrays__ have been widely used in human population genetics studies, this was not what they were designed for. The SNP arrays are a useful tool for studying slight variations between whole genomes, and their most important clinical applications are for determining disease susceptibility and for measuring the efficacy of drug therapies designed specifically for individuals. In research, SNP arrays are most frequently used for genome-wide association studies.

The ascertainment and choice of the SNPs are not well documented, and the tagging design is not ideal for population studies (__ascertainment bias__). To address this problem, the __Affymetrix Axiom Human Origins Array__ has been designed. This analyzes ~630,000 SNPs ascertained by resequencing in 13 population-specific panels, including San, Yoruba, Mbuti Pygmy, French, Sardinian, Han, Cambodian, Mongolian, Karitiana, Papuan, and Bougainville populations from the CEPH-HGDP panel, as well as variants identified from the Neanderthal Genome Project. The array also contains 87,000 SNPs that include sets from mtDNA and the Y chromosome, and variants from standard chips for data comparison purposes.

Dataset 11 contains __629,443 SNPs__ that were obtained by genotyping __934 unrelated HGDP-CEPH individuals__ with Affymetrix Axiom Human Origins Array Plate, and merging the genotypes with data from _Neandertal_, _Denisova_ and _chimpanzee_. 

The SNP data are divided among 14 partially overlapping datasets, 13 of which are of value for analysis of different population genetics scenarios. 

For each of __datasets 1-12__, SNPs were discovered as heterozygotes by whole-genome shotgun sequencing of a different HGDP-CEPH individual of known ancestry, as per [Keinan et al. 2007](https://www.nature.com/articles/ng2116). 

__Dataset 13__ contains heterozygote SNPs for each of which a random _Denisovan_ allele matches that of _chimpanzee_, and the random _San_ allele is derived. 

__Dataset 14__, which is valuable for studying population structure using a maximum number of SNPs, _does not allow demographic modeling_. This dataset combines __all SNPs__ along with an __additional 87,044 SNPs chosen to allow haplotype inference at mitochondrial DNA and the Y chromosome__, and to provide overlap with previous Affymetrix and Illumina genotyping arrays so that users can merge the data available here with previously published datasets. 

Read the [detailed technical document](ftp://ftp.cephb.fr/hgdp_supp10/8_12_2011_Technical_Array_Design_Document.pdf) before using these data in order to avoid pitfalls. The array was developed by David Reich and colleagues in collaboration with Affymetrix for the purpose of generating data with clearly documented ascertainment.

All the data that we need for the exercises are located here:
`/opt/evop/public/BIOINFORMATICS/popgen_intro/`

We will copy them to our home directory instead of downloading them so we can save some time.
```{bash, eval=F, echo=T}
mkdir ~/popgen_intro
cd /opt/evop/public/BIOINFORMATICS/popgen_intro/
cp -r HGDP_metainformation.txt Harvard_HGDP-CEPH/all_snp.* Harvard_HGDP-CEPH/sample_all_snp.txt ~/popgen_intro/
cd ~/popgen_intro
```




We can download the dataset like this:
```{bash, eval=F, echo=T}
mkdir ~/popgen_intro
cd ~/popgen_intro
wget ftp://ftp.cephb.fr/hgdp_supp10/*
```

A tarball is a group of files that are kept together using the `tar` command. Tarballs are common file formats on Linux operating systems, and they are often used for distribution of software/media or backup purposes. Typically they have `.tar` extension, while compressed `.tar` files have `.tgz` or `.tar.gz` extension.

```{bash, eval=F, echo=T}
tar -xzvf Harvard_HGDP-CEPH.tgz
```

`Harvard_HGDP-CEPH` directory which contains several files explained in `8_12_2011_Harvard_HGDP_readme.txt`. 

in `Harvard_HGDP-CEPH`  we can find several files:<br />
`all_snp.ped`<br />
`all_snp.map`<br />

These files are in the `plink` format (used by software such as `ADMIXTURE`, `TreeMix`, and many others). These two files are "connected" and they should be always processed together. So be sure that you keep them in the same directory and that they have the same name with only difference in their file extension name.

`sample_all_snp.txt`<br /> contains a list of individuals their sex and populations to which they belong<br />

`HGDP_metainformation.txt` is a file in which I collected the meta-information on each of the samples which might be useful for population genetic research, data interpretation, and/or data visualization (e.g. country or origin, longitude, latitude, language...).<br />

For all data manipulation, such as extracting certain SNPs, chromosomes, individuals, merging datasets, removing SNPs in high LD, filtering out low quality genotypes, and many others we will use [plink](https://www.cog-genomics.org/plink2).

*********************************************************************

<br /><br /><br />

# Plink

We will be working with `plink`, a free, open-source whole-genome association analysis toolset, designed to perform a range of basic, large-scale analyses in a computationally efficient manner. Even though the focus of `plink` is  on analysis of genotype/phenotype data, it is widely used in popgen as it has many features for data manipulation, it offers basic statistics, and many popgen tools assume input files to be in plink format (e.g. `ADMIXTURE`).

[Plink](https://www.cog-genomics.org/plink/1.9/) is a tool created for genome-wide association studies (GWAS) and research in population genetics. `Plink` parses each command line as a collection of flags (each of which starts with two dashes), plus parameters (which immediately follow a flag). Because Plink was developed for GWAS medical studies, many statistics and pieces of information from plink files will be not used in our analysis, such as pedigree or phenotype.

*********************************************************************

## Plink file formats

`plink` can either read 
<br />1. __text-format files__ (`.ped` + `.map`) or 
<br />2. __binary files__ (`.bed` + `.bim` + `.fam`). 

Because reading large text files can be time-consuming, it is recommended to use binary files. 

![plink data formats (Marees et al. 2017)](C:\Users/Vladimir Bajic/Documents/Nowick_Lab/Master_course_Bioinformatics_for_Biologists/SoSe2020/popge_intro/figures_for_html/PLINK-files.png){width=100%}

*********************************************************************

### Text-format plink files

Text `plink` data consist of two files (which should have matching names and they should be stored together): 
<br />1. `.ped` contains information on the __individuals and their genotypes__; 
<br />2. `.map` contains information on the __genetic markers__.

*********************************************************************

### Binary plink files

Binary `plink` data consist of three files (one binary and two text files which should have matching names and which should be stored together):
<br />1. `.bed`  contains individual identifiers (IDs) and genotypes, 
<br />2. `.fam` contain information on the individuals, 
<br />3. `.bim` contains information on the genetic markers. 

Analysis using __covariates__ often requires the fourth file, containing the values of these covariates for each individual.

*********************************************************************

## Basic statistics with plink

We can generate some simple summary statistics using `plink` e.g. rates of missing data in the file, diversity within and between the samples.

Before we start our exercises let's make binary files out of our text-format plink files and output them in `~/popgen_intro/plink_exercise/`:
```{bash, echo=T, eval=F}
cd ~/popgen_intro
plink1.9 --file all_snp --make-bed --out ~/popgen_intro/plink_exercise/hgdp
```

```{bash, echo=T, eval=F}
PLINK v1.90b6.6 64-bit (12 Oct 2018)           www.cog-genomics.org/plink/1.9/
(C) 2005-2018 Shaun Purcell, Christopher Chang   GNU General Public License v3
Logging to /home/bajiv90/popgen_intro/plink_exercise/hgdp.log.
Options in effect:
  --file all_snp
  --make-bed
  --out /home/bajiv90/popgen_intro/plink_exercise/hgdp

16042 MB RAM detected; reserving 8021 MB for main workspace.
.ped scan complete (for binary autoconversion).
Performing single-pass .bed write (627719 variants, 942 people).
--file: /home/bajiv90/popgen_intro/plink_exercise/hgdp-temporary.bed +
/home/bajiv90/popgen_intro/plink_exercise/hgdp-temporary.bim +
/home/bajiv90/popgen_intro/plink_exercise/hgdp-temporary.fam written.
627719 variants loaded from .bim file.
942 people (621 males, 321 females) loaded from .fam.
Using 1 thread (no multithreaded calculations invoked).
Before main variant filters, 942 founders and 0 nonfounders present.
Calculating allele frequencies... done.
Warning: 3 het. haploid genotypes present 
(see /home/bajiv90/popgen_intro/plink_exercise/hgdp.hh ); 
many commands treat these as missing.
Warning: Nonmissing nonmale Y chromosome genotype(s) present; 
many commands treat these as missing.
Total genotyping rate is 0.995272.
627719 variants and 942 people pass filters and QC.
Note: No phenotypes present.
--make-bed to /home/bajiv90/popgen_intro/plink_exercise/hgdp.bed +
/home/bajiv90/popgen_intro/plink_exercise/hgdp.bim +
/home/bajiv90/popgen_intro/plink_exercise/hgdp.fam ... done.
```

Now go to `plink_exercise` and check which files are there:

```{bash, echo=T, eval=F}
cd ~/popgen_intro/plink_exercise/
ls 
```


```{bash, echo=T, eval=F}
hgdp.bed  hgdp.bim  hgdp.fam  hgdp.hh  hgdp.log
```

As expected we got 3 files that go together in binary file format in `plink` (i.e. `bed` + `bim` + `fam`), but also two more files `hh` and `log`. Let's take a look at them:

```{bash, echo=T, eval=F}
head hgdp.hh hgdp.log
```

```{bash, echo=T, eval=F}
==> hgdp.hh <==
HGDP00199       HGDP00199       Affx-50065401
HGDP00259       HGDP00259       Affx-50065401
HGDP00281       HGDP00281       Affx-50065401

==> hgdp.log <==
PLINK v1.90b6.6 64-bit (12 Oct 2018)
Options in effect:
  --file all_snp
  --make-bed
  --out /home/bajiv90/popgen_intro/plink_exercise/hgdp

Hostname: evop-login
Working directory: /home/bajiv90/popgen_intro/
Start time: Thu Apr 30 11:06:36 2020

Random number seed: 1588237596
16042 MB RAM detected; reserving 8021 MB for main workspace.
Scanning .ped file... done.
Performing single-pass .bed write (627719 variants, 942 people).
--file: /home/bajiv90/popgen_intro/plink_exercise/hgdp-temporary.bed +
/home/bajiv90/popgen_intro/plink_exercise/hgdp-temporary.bim +
/home/bajiv90/popgen_intro/plink_exercise/hgdp-temporary.fam written.
627719 variants loaded from .bim file.
942 people (621 males, 321 females) loaded from .fam.
Using 1 thread (no multithreaded calculations invoked).
```

`plink` created `.hh` file to tell us that there is an issue with heterozygous haploid and nonmale Y chromosome genotype calls. If we take a better look at `.log` file we will see that there is a short explanation for files that were created, so for `hh` we can find this: 
```{bash, echo=T, eval=F}
Warning: 3 het. haploid genotypes present 
(see /home/bajiv90/popgen_intro/plink_exercise/hgdp.hh ); 
many commands treat these as missing.
Warning: Nonmissing nonmale Y chromosome genotype(s) present; 
many commands treat these as missing.
```
In our case, there are only 3 SNPs in total that are problematic and we can just exclude them from further analysis if we would be interested in Y chromosome. But we will focus our analysis on autosomes and X chromosome, so no need to bother with those for now. See [plink webpage](https://www.cog-genomics.org/plink/1.9/data) for more ideas on how to deal with issues like this. 


We can include population names as family ID (FID) so we can have an idea where to which population individuals belong. We can get this information from the metainformation file that we have in `~/popgen_intro/HGDP_metainformation.txt`. To do so we should first make a file with old and new IID and FID, and then making new plink files with updated IDs:

```{bash, eval=F, echo=T}
cd ~/popgen_intro/plink_exercise/

join -j 1 -o 1.1,1.2,2.2,2.1 <(sort -k1 hgdp.fam) <(sort -k1 ../HGDP_metainformation.txt) > hgdp_updateIDs.txt

plink1.9 --bfile hgdp --update-ids hgdp_updateIDs.txt --make-bed --out hgdp_newFIDs

```



<div class="alert alert-info" role="alert">
  <strong>TIP: Process Substitution</strong>
  
  *****
__Piping__ (`|`) the stdout of a command into the stdin of another is a powerful technique. But, what if we need to pipe the stdout of multiple commands? This is where __process substitution__ (`<(command)` or `>(command)`) comes in.

Process substitution feeds the output of a process (or processes) into the stdin of another process. In other words process substitution treats the output (or input) of commands as files. 
`>(command list)`  will treat it as an input
`<(command list)`  will treat it as an output

The syntax for process substitution is: `<(list)` or `>(list)`
e.g. `diff -u <( ls dir1 | sort)  <( ls dir2 | sort )`
</div>



***************************************************************************

### Missing data

We can use `--missing` to produces __sample-based__ (`.imiss`) and __variant-based__ (`lmiss`) missing data reports.
```{bash, echo=T, eval=F}
cd ~/popgen_intro/plink_exercise/

plink1.9 --bfile hgdp_newFIDs --missing --out hgdp_newFIDs_miss
```

Now, we can visualize the data in R. Let's download newly created files together with the metainformation file to our local machine, and then run the following R commands to plot them:

```{r plot_miss, eval=T, echo=T, fig.width=20, fig.align='center'}

# Loading necessary packages
library(data.table)
library(dplyr)

# Specifying path to directory and loading files
setwd("~/Nowick_Lab/Master_course_Bioinformatics_for_Biologists/SoSe2020/popge_intro/")
indINFO <- fread("HGDP_metainformation.txt")
imiss <- fread("hgdp_newFIDs_miss.imiss")
lmiss <- fread("hgdp_newFIDs_miss.lmiss")

# to vlookup to indINFO table 
imiss <- left_join(imiss, indINFO, by="IID")

# Preparing to plot Missingness boxplots per population
# First ordering the populations based on region and color so that they appear like that on the boxplot
ordered_imiss <- imiss[order(imiss$Region, imiss$Country, imiss$FID),]
imiss$FID <- factor(imiss$FID, levels = unique(ordered_imiss$FID))

# ploting boxplots per population colored by region
library("ggplot2")
ggplot(imiss, aes(x=FID, y=F_MISS, fill=Region)) + 
  geom_boxplot() + 
  theme(axis.text.x = element_text(angle = 90))

# Let's keep only modern human populations
imiss_sub <- filter(imiss, !FID %in% c("Href","Clint","Denisova", "Gorilla","Macaque","Marmoset","Orang","Vindija"))

# and plot it again so we can see better which of modern populations has a lot of individuals that are missing a lot of data
ggplot(imiss_sub, aes(x=FID, y=F_MISS, fill=Region)) + 
  geom_boxplot() + 
  theme(axis.text.x = element_text(angle = 90))

# Preparing to plot histograms of Missingness per Chromosome
# First we will reorder factors so chromosomes on the x-axis are displayed in proper order
library(forcats)
lmiss$CHR <- as.factor(lmiss$CHR)
lmiss$CHR <- fct_relevel(lmiss$CHR, function(x){as.character(sort(as.integer(x)))})

ggplot(lmiss, aes(x=CHR, y=F_MISS, fill=CHR)) + 
  geom_boxplot() +
  theme(legend.position = "none")

# Now let's see only those that have F_MISS < 0.1
lmiss_sub <- filter(lmiss, F_MISS < 0.1)
lmiss_sub$CHR <- as.factor(lmiss_sub$CHR)

ggplot(lmiss_sub, aes(x=CHR, y=F_MISS, fill=CHR)) + 
  geom_boxplot() +
  theme(legend.position = "none")
  
```


Now, let's keep only individuals from modern populations, and remove SNPs and individuals with more than 10% missingness. To do so, we will need to specify which individuals we want to keep (with option `--keep`) or remove (with option `--remove`). One way to make list of individuals which we want to keep is to find all individuals that contain "HGDP" in their name.
```{bash, echo=T, eval=F}
grep HGDP hgdp_newFIDs.fam > ind_keep.txt
plink1.9 --bfile hgdp_newFIDs --make-bed --mind 0.1 --geno 0.1 --keep ind_keep.txt --out hgdp_newFIDs_modernPops_MindGeno
```

How many individuals and how many SNPs were removed?


***************************************************************************


### IBD and PI_HAT

![Consanguinity - degrees of relationships](C:\Users/Vladimir Bajic/Documents/Nowick_Lab/Master_course_Bioinformatics_for_Biologists/SoSe2020/popge_intro/figures_for_html/relatives.jpg){width=100%}

The HGDP data set is already filtered for relatives according to [Rosenberg et al. 2006](https://www.ncbi.nlm.nih.gov/pubmed/17044859). There are different ways to determine related individuals. Here we will mention plink's widely used __IBD__ estimate __PI_HAT__ ($\hat{\pi}$) which can be used to estimate the level of relatedness between a pair of individuals. PI_HAT is a measure of overall IBD alleles. It measures the fraction of the genome shared (IBD) between pairs of the individuals (P(IBD=2)+0.5*P(IBD=1)). 

Since our data is already filtered for relatives using more sophisticated methodologies (see Rosenberg et al. 2006) we will only use PI_HAT to search for individuals that still have very high PI_HAT. This will illustrate to us how isolation, consanguinity, bottleneck, drift can violate simple principles on which PI_HAT relies on (e.g. populations from Oceania and Americas).

```{bash, echo=T, eval=F}
plink1.9 --bfile hgdp_newFIDs_modernPops_MindGeno --genome --out hgdp_newFIDs_modernPops_MindGeno_genome
```

Check plink webpage for more information about the [output of the `--genome` option](https://www.cog-genomics.org/plink/1.9/formats#genome).

![Degrees of relationships](C:\Users/Vladimir Bajic/Documents/Nowick_Lab/Master_course_Bioinformatics_for_Biologists/SoSe2020/popge_intro/figures_for_html/IBD.png){width=100%}

Columns Z0, Z1, and Z2 indicate the probabilities of having IBD of 0, 1, or 2 over the loci, which gives us a way of discriminating between relationship types. 
Ideal parent-offspring has (Z0, Z1, Z2) = (0, 1, 0), i.e. all loci have one allele identical by descent; 
Ideal full sibling = (1/4, 1/2, 1/4), i.e. 25% of loci have 0 alleles IBD, 50% have 1 allele IBD, 25% have 2 alleles IBD; etc. So, we can use Z0, Z1, and Z2 to distinguishing between relationship types. 

IBD=0, IBD=1, and IBD=2 are the genome proportions shared on 0, 1, and 2 chromosomes, respectively, between two individuals. Many relationships share the same expected mean IBD proportions; however, for full-sibling, second-degree, and third-degree relationships, a variance around the expected mean is due to the random nature of recombination events. Genotyping and other technical errors can contribute to this variance.

The table below shows the expected mean IBD proportions for the outbred familial relationship categories

<center>

| Familial Relationship |	IBD0 |	IBD1 | IBD2 |
|-----------------------|------|-------|------|
|Parental               |	0    |	1    |	0   |
|Full-sibling           |	0.25 |	0.5  |	0.25|
|Half-sibling, avuncular, and grandparental |	0.5 |	0.5 |	0 |
|First-cousin, great-grandparental, great-avuncular, and half-avuncular |	0.75 |	0.25 |	0 |
|Distantly related      |	varies |	varies |	0 |
|Unrelated (includes relationships beyond the third degree) |	1 |	0 |	0 |

</center>

In theory, PI_HAT values that are around: 
<br />0.125 are __third degree__, 
<br />0.25 are __second degree__, and 
<br />0.5 are __first degree__. 
<br />


PI_HAT assumes homogeneous population with known population allele frequencies, however if the population is not homogeneous, the values are biased compared to the expectations under the known relationship categories. 

In other words, plink's PI_HAT is based on the assumption that the study sample is drawn from a single, homogeneous, randomly mating population. This assumption is violated if pedigree founders are drawn from multiple populations or include admixed individuals. In the presence of population structure, the method of moments estimator has an inflated variance and can be biased because it relies on sample-based allele frequency estimates. For plink's estimator which truncates genome-wide sharing estimates at zero and one to generate biologically interpretable results, the bias is most often towards over-estimation of relatedness between ancestrally similar individuals. 

Thus the PI_HAT/Z0/Z1/Z2 values produced by `plink --genome` should not be taken as solid evidence to justify relationship inference. Instead it is better to use [KING](http://people.virginia.edu/~wc9c/KING/) or [PC-Relate](https://www.rdocumentation.org/packages/GENESIS/versions/2.2.2/topics/pcrelate).


*******************************************************

### Heterozygosity - inbreeding - consanguinity

A __method-of-moments estimator__ (__F__) developed by Ritland (1996) can be used to estimate the distribution of __inbreeding__ among and __relatedness__ between individuals of a natural population.

Plink's `--het` option computes observed and expected autosomal homozygous genotype counts for each sample, and reports method-of-moments F coefficient estimates (i.e. ([observed hom. count] - [expected count]) / ([total observations] - [expected count])) to `plink.het`.

```
plink1.9 --bfile hgdp_newFIDs_modernPops_MindGeno --het --out hgdp_newFIDs_modernPops_MindGeno_het
```

We can visualize the difference in heterozygosity within populations and we can visualize it on the map like this:
```{r, echo=T, eval=T, fig.align='center'}
# Loading necessary packages
library(data.table)
library(dplyr)
library("ggplot2")

# Specifying path to directory and loading files
setwd("~/Nowick_Lab/Master_course_Bioinformatics_for_Biologists/SoSe2020/popge_intro/")
indINFO <- fread("HGDP_metainformation.txt")
het <- fread("hgdp_newFIDs_modernPops_MindGeno_het.het")


#to vlookup to indINFO table 
het <- left_join(het, indINFO, by="IID")

# Preparing to plot Missingness boxplots per population
# First ordering the populations based on region and color so that they appear like that on the boxplot
ordered_het <- het[order(het$Region, het$Country, het$FID),]
het$FID <- factor(het$FID, levels = unique(ordered_het$FID))

# ploting boxplots per population colored by region

ggplot(het, aes(x=FID, y=F, fill=Region)) + 
  geom_boxplot() + 
  theme(axis.text.x = element_text(angle = 90))


# Calculate mean F per population
pop_mean_F =  aggregate(F ~ FID, het, mean )

# Extract metainformation for each population and merge it with pop_mean_F 
uniqMetaPop <- unique(het[,c("FID", "Region", "Latitude", "Longitude")])
pop_mean_F_extra = left_join(pop_mean_F, uniqMetaPop, by="FID")


library("maps")
world_map <- map_data(map="world")

ggplot() + 
  theme(panel.background = element_rect(fill = "white")) +
  geom_map(data=world_map, map=world_map, 
           aes(map_id=region), fill="black", colour="gray20", size=0.15) +
  coord_quickmap(ylim=c(-35,70), xlim=c(-115,150)) +
  geom_point(data=pop_mean_F_extra, aes(x=Longitude, y=Latitude, color=F), alpha =0.5, size=5 )+
  scale_color_gradientn(colours = rainbow(10)) +
  ggtitle("Average method-of-moments [F] coefficient estimates for HGDP populations")
```

In which populations F has lowest and highest values? How would you interpret these results?


****************************************************************************************************


<br /><br /><br />

# Multivariate analyses

_How can we display a set of pairwise genetic distances in a comprehensible manner?_ 
If we have n populations, we require n dimensions to fully display their pairwise genetic distances on a graph. However, we often have more than four populations or molecules that we want to compare, yet we cannot conceive of, or represent, the four or more dimensions required to display these data. __Multivariate analyses__ allow us to reduce these multiple dimensions to the two or three dimensions we can draw and comprehend, while minimizing the inevitable loss of information. 

![How many dimensions are needed to display population relationships?](C:\Users/Vladimir Bajic/Documents/Nowick_Lab/Master_course_Bioinformatics_for_Biologists/SoSe2020/popge_intro/figures_for_html/mds_example.jpg){width=100%}

Figure from [Jobling et al. 2014](https://www.amazon.com/Human-Evolutionary-Genetics-Mark-Jobling/dp/081534148)

Distance matrices are shown relating genetic distances between increasing numbers of populations. As the number of populations increases, the number of dimensions needed to display these distances visually such that the lengths of the lines represent the genetic distances between them quantitatively also increases. A single line of a given length can relate two populations. Three populations can be related by a triangle, whose sides have lengths proportional to the genetic distance. Four populations can be represented by a pyramid. With five or more populations it is no longer possible to represent population relationships in three-dimensional space.


****************************************************************************************************

## MDS
__Multi-dimensional scaling (MDS)__ is one of these methods, where the loss of information is represented as a __stress statistic__, with the lower the stress statistic, the better the MDS fit to the data and therefore the less information lost. 

We can use plink to construct MDS for `.genome` file that we created
```{bash, echo=T, eval=F}
# First we can create 4 dimensions for MDS
plink1.9 --bfile hgdp_newFIDs_modernPops_MindGeno --read-genome hgdp_newFIDs_modernPops_MindGeno_genome.genome --cluster --mds-plot 4 --out mds_4D_hgdp


# And then we can calculate MDS with 2 dimensions
plink1.9 --bfile hgdp_newFIDs_modernPops_MindGeno --read-genome hgdp_newFIDs_modernPops_MindGeno_genome.genome --cluster --mds-plot 2 --out mds_2D_hgdp
```



```{r, MDS_plot, eval=T, echo=T, fig.width=15, fig.align='center'}

# MDS plot

# Loading necessary packages
library(data.table)
library(dplyr)
library("ggplot2")

# Specifying path to directory and loading files
setwd("~/Nowick_Lab/Master_course_Bioinformatics_for_Biologists/SoSe2020/popge_intro/")
indINFO <- fread("HGDP_metainformation.txt")
mds <- fread("mds_2D_hgdp.mds")
#mds <- fread("mds_4D_hgdp.mds") # try with 4 dimensions

# to vlookup to indINFO table 
mds_inf <- left_join(mds, indINFO, by="IID")

# Deffine shape for each population within each Region
symbol_col <- unique(mds_inf[,c("FID","Region")])
symbol_col$Symbol <- NA
symbol_col$Color <- NA

# assign shape/symbol
for (i in unique(symbol_col$Region)){
  symbol_col[which(symbol_col$Region == i),"Symbol"] <- 1:nrow(symbol_col[which(symbol_col$Region == i),])
}

# assign color
n = 0
for (i in unique(symbol_col$Region)){
  n = n + 1
  symbol_col[which(symbol_col$Region == i),"Color"] <- n
}

# Merge with mds dataframe
mds_inf_shape <- left_join(mds_inf, symbol_col[,c("FID","Symbol","Color")], by="FID")
mds_inf_shape$Region_Population <- paste(mds_inf_shape$Region, mds_inf_shape$Population, sep = " - ")

# Order dataset before plotting
ordered_mds_inf_shape <- mds_inf_shape[order(mds_inf_shape$Region_Population),]
# MDS plot with colors according to Regions and shapes as Populations
ggplot(mds_inf_shape, aes(x=C1, y=C2)) +
  geom_point(aes(col=Region_Population, shape=Region_Population)) +
  scale_colour_manual(values = c("#a6cee3","#1f78b4", "#b2df8a", "#33a02c","#fb9a99", "#e31a1c", "#fdbf6f")[as.factor(unique(ordered_mds_inf_shape[,c("Color","Region_Population")])$Color)]) +
  scale_shape_manual(values=unique(ordered_mds_inf_shape[,c("Symbol","Region_Population")])$Symbol) +
  theme_bw()

```



****************************************************************************************************

## PCA

It is also possible to use multivariate analysis to generate two- or three-dimensional graphical representations of distances between populations using the _raw data of allele frequencies_, rather than _genetic distances_. __Principal component analysis (PCA)__ is a commonly used example of this approach, and represents one of the most useful techniques to visualise genetic diversity in a dataset. The methodology is not restricted to genetic data, but in general allows breaking down high-dimensional datasets to two or more dimensions for visualization in a two-dimensional space. Individual axes, known as __principal components (PCs)__ or __eigenvectors__, are extracted sequentially, with each PC being independent and encapsulating as much of the remaining variation as possible. Using PCA it is possible to estimate the proportion of the total variance in the dataset that has been summarized within these reduced dimensions. PCA can reveal global relationships from a set of _pairwise distances between populations_.

PCA is also used to display genetic distances between a __set of individuals__, rather than a __set of populations__. Each data point represents an individual person, plotted according to the eigenvalues of the first two PCs of the data calculated from the __covariance matrix__ of a genome-wide SNP dataset. Plotting values for two PCs for every individual allows the spatial separation of the different groups.

For the plots of genome-wide data, the first two PCs are normally shown, but any two PCs can be plotted to search for useful information. It should be remembered that, for genome-wide analysis, although the first two PCs will represent the largest amount of variation of all PCs, they may still represent a small amount of the total variation in the dataset. Principal components can also be rotated, a statistical procedure that is sometimes used to emphasize the relationship between the genetic structure and geography.

![Population structure within Europe](C:\Users/Vladimir Bajic/Documents/Nowick_Lab/Master_course_Bioinformatics_for_Biologists/SoSe2020/popge_intro/figures_for_html/PCA_Europe.jpg){width=100%}

Population structure within Europe revealed by genome-wide SNPs. Principal component analysis of data on 197,146 SNPs in 1387 Europeans. Small colored labels represent individuals and large colored points represent median PC1 and PC2 values for each country. The inset map provides a key to the labels. The PC axes are rotated to emphasize the similarity to the geographic map of Europe (from [Novembre J et al. 2008](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2735096/)).

Soon we will see that PC plots based on genome-wide SNP analyses show that world populations from the HGDP-CEPH collection cluster according to the continent, and as shown here European populations cluster approximately according to country. This indicates that our genomes contain detailed information about geographical ancestry, a finding with important implications for __persona genomics__ and __forensic genetics__.

****************************************************************************************************


### PCA in plink

With the modifier `--pca` PLINK extracts the top 20 principal components of the variance-standardized relationship matrix. The results consist in a `.eigenvec` file with the coordinates for each individual in rows and eigenvectors in columns, and a `.eigenval` which explains how much variance there is in the data for that given vector.

```{bash plink_PCA, echo=T, eval=F}
plink1.9 --bfile hgdp_newFIDs_modernPops_MindGeno --pca --out hgdp_newFIDs_modernPops_MindGeno_PCA
```


Now we can visualize PCA in R:
```{r PCA_plot, echo=T, eval=T, fig.width=10, fig.align='center'}

# PCA

# Loading necessary packages
library(data.table)
library(dplyr)
library("ggplot2")

# Specifying path to directory and loading files
setwd("~/Nowick_Lab/Master_course_Bioinformatics_for_Biologists/SoSe2020/popge_intro/")

indINFO <- fread("HGDP_metainformation.txt")

eigenvec <- fread("hgdp_newFIDs_modernPops_MindGeno_PCA.eigenvec")
names(eigenvec) <- c("FID","IID", paste("PCA",1:20, sep = "_"))

eigenval <- fread("hgdp_newFIDs_modernPops_MindGeno_PCA.eigenval")
eigneval <- data.frame("PCA" = 1:20, "Eigenval" = eigenval$V1)

# Plot eigenval
ggplot(data=eigneval, aes(x=PCA, y=Eigenval_prop)) +
  geom_bar(stat="identity")

# Or we can calculate proportion for each eigenval from total eigenval and plot it
eigneval$Eigenval_prop <- round(eigneval$Eigenval / sum(eigneval$Eigenval)*100,2)

ggplot(data=eigneval, aes(x=PCA, y=Eigenval_prop)) +
  geom_bar(stat="identity")


# Preparing to plot PCA
# to vlookup to indINFO table 
pca_inf <- left_join(eigenvec, indINFO, by="IID")

# Deffine shape for each population within each Region
symbol_col <- unique(pca_inf[,c("FID","Region")])
symbol_col$Symbol <- NA
symbol_col$Color <- NA

# assign shape/symbol
for (i in unique(symbol_col$Region)){
  symbol_col[which(symbol_col$Region == i),"Symbol"] <- 1:nrow(symbol_col[which(symbol_col$Region == i),])
}

# assign color
n = 0
for (i in unique(symbol_col$Region)){
  n = n + 1
  symbol_col[which(symbol_col$Region == i),"Color"] <- n
}

# Merge with mds dataframe
pca_inf_shape <- left_join(pca_inf, symbol_col[,c("FID","Symbol","Color")], by="FID")
pca_inf_shape$Region_Population <- paste(pca_inf_shape$Region, pca_inf_shape$Population, sep = " - ")

# Order dataset before plotting
ordered_pca_inf_shape <- pca_inf_shape[order(pca_inf_shape$Region_Population),]
# PCA plot with colors according to Regions and shapes as Populations
ggplot(pca_inf_shape, aes(x=PCA_1, y=PCA_2)) +
  geom_point(aes(col=Region_Population, shape=Region_Population)) +
  scale_colour_manual(values = c("#a6cee3","#1f78b4", "#b2df8a", "#33a02c","#fb9a99", "#e31a1c", "#fdbf6f")[as.factor(unique(ordered_pca_inf_shape[,c("Color","Region_Population")])$Color)]) +
  scale_shape_manual(values=unique(ordered_pca_inf_shape[,c("Symbol","Region_Population")])$Symbol) +
  theme_bw()
```

Do you notice any outliers? Let's check other PCA's and figure out which individuals should be removed. 
As this is well established dataset, there is already indication in `~/popgen_intro/sample_all_snp.txt`


<div class="alert alert-warning">
  <strong>HOMEWORK:</strong> 
  
  *****
  Make a new PCA after removing the outliers.

  1. Make file with individuals to be removed;
     
  2. Use this file to subset your original file;
     
  3. Run PCA on new file that doesn't contain outlier individuals; 
     
  4. Visualize PCA in R.
  
  <br />Do you see any differences?
  
  *****************
  
  <details>
  <summary><strong>Click for Answer</strong></summary>
  <br />1. Make file with individuals to be removed;
     
```{bash, eval=F, echo=T}
cd ~/popgen_intro/plink_exercise
grep discover ~/popgen_intro/sample_all_snp.txt > ind_exclude_tmp.txt
join -j 1 -o 2.2,1.1 <(sort -k1 ind_exclude.txt) <(sort -k1 ../HGDP_metainformation.txt) > ind_exclude.txt
```
     
  <br />2. Use this file to subset your original file;
     
```{bash, eval=F, echo=T}
plink1.9 --bfile hgdp_newFIDs_modernPops_MindGeno --remove ind_exclude.txt --make-bed --out hgdp_outliers_removed
```
     
  <br />3. Run PCA on new file that doesn't contain outlier individuals; 
     
```{bash, eval=F, echo=T}
plink1.9 --bfile hgdp_outliers_removed --pca --out hgdp_outliers_removed_PCA
```
     
  <br />4. Download necessary ouput files to your local machine and visualize PCA in R using the script abowe. Don't forget to exchange input files for `eigenval` and `eigenvec`.
  
  </details> 

</div>

***************************************************************************




<br /><br /><br />

# ADMIXTURE



[ADMIXTURE](https://www.genetics.ucla.edu/software/admixture/) is a software that we can use to identify ancestry components shared between individuals in a set of populations. It takes `plink` format input files.


*********************************************************************

### Pruning

Before running ADMIXTURE, it is advised to prune the dataset. We will use `plink` to prune the dataset by excluding the SNPs which are in Linkage. The resulting file will have fewer SNPs, and the computation will be faster. We will also only keep SNPs that are on the autosomes (`--autosome`)

<br />To prune for LD we will use these settings previously recommended for HO array `--indep-pairwise 200 25 0.4`:
<br />window size in SNPs = 200;
<br />number of SNPs to shift the window at each step = 25;
<br />r^2 threshold (pairwise SNP-SNP metric) = 0.4;

<br />Using these parameters specified above we will 
<br />a) consider a window of 200 SNPs,
<br />b) calculate LD between each pair of SNPs in the window,
<br />b) remove one of a pair of SNPs if the LD is greater than 0.4,
<br />c) shift the window 25 SNPs forward and repeat the procedure.


```{bash, echo=T, eval=F}
plink1.9 --bfile hgdp_newFIDs_modernPops_MindGeno --indep-pairwise 200 25 0.4 --out prune_hgdp
plink1.9 --bfile hgdp_newFIDs_modernPops_MindGeno --extract prune_hgdp.prune.in --recode12 --make-bed --autosome --out hgdp_pruned_autosome
```

How many SNPs are left after pruning?

*****************************************************

ADMIXTURE demands a bit more computational power and time than previous analysis we did, and we will not be able to run it on `evop-login` server. So, instead I will explain how to run ADMIXTURE and then we will focus on visualizing the output of the ADMIXTURE in R. 


*************************************************************

### Running ADMIXTURE

We can use the following commands to run ADMIXTURE for each K. One of the K values will have better supported than the others and we can find it by searching for the one K which has the lowest __cross-validation (cv) error__. And finaly, we will use the best K as the best representation of the actual data variation and visualize it.

```
for K in 6 7 8 9 
do
  admixture -s $RANDOM --cv hgdp_pruned_autosome.ped $K | tee log.K${K}.out;
done
```

This will create three output files for each K: 
<br />1.`.out` 
<br />1.`.P`
<br />1.`.Q`

*************************************************************

### Cross-validation error to identify the best value of K

We can find the best supported K by looking at the `.out` files in a section where is CV error reported. We can take a look at the file with command `less` and then we will see that we can simply grep pattern "CV" from all our log files and store in new textfile like this:

```{bash, eval=F, echo=T}
grep -h CV log*out | awk -F'[(|=|)|\t| ]' '{print $5 "\t" $7}' > CV.txt
```

Now we can download thsi file to our local machines and plot it in R.
```{r, CV_plot, eval=T, echo=T, fig.align='center'}
library(data.table)
library(dplyr)
library("ggplot2")

setwd("~/Nowick_Lab/Master_course_Bioinformatics_for_Biologists/SoSe2020/popge_intro/")
cv <- fread("CV.txt", col.names = c("K","cv"))   

ggplot(cv, aes(group=1, x=K, y=cv)) + 
  geom_point() +
  geom_line() +
  ggtitle("CV error for each K")
```

When performing ADMIXTURE we should also do several runs for each K. For simplicity reasons in our example we did only one run for each K. When multiple runs are present for each K we can determine which one of them has the highest likelihood, and then we can use it as the representative run for the given K.

*************************************************************

### Plotting ADMIXTURE results

For simplicity reasons, we will plot the results of the admixture on chromosome 6. In reality, one would run admixture on all autosomal chromosomes for multiple K's and with multiple runs for each K.

First copy files, and download them to your local machine, and then visualize results in R.
```{bash, eval=F, echo=T}
cp -r /opt/evop/public/BIOINFORMATICS/popgen_intro/files_4_admixture ~/popgen_intro/
```

And now once you downloaded the files you can plot the results in R:

```{r ADMIXTURE_plot, eval=T, echo=T, fig.width=30, fig.align='center'}
library(data.table)
library(dplyr)
library(ggplot2)
library(tidyr)

setwd("~/Nowick_Lab/Master_course_Bioinformatics_for_Biologists/SoSe2020/popge_intro/")

# read in the necessary files
q_file <- read.table("hgdp_pruned_chr6.9.Q")
fam_file <- read.table("hgdp_pruned_chr6.fam", col.names = c("Population", "IID", "PID","MID","Sex","P"))
popGroups <- read.table("HGDP_metainformation.txt", header = T)

# Prepare the data by adding new columns that contain information abut Region and Population
q_file_ID <- cbind(fam_file[,1:2], q_file)
q_file_ID_Region <- left_join(q_file_ID, popGroups[,c("Population","Region")], by="Population")

# Tranforme the data before plotting
plot_data <- q_file_ID_Region %>% 
  mutate(id = row_number()) %>% 
  gather('pop', 'prob', V1:V9) %>% 
  group_by(id) %>% 
  mutate(likely_assignment = pop[which.max(prob)],
         assingment_prob = max(prob)) %>% 
  arrange(likely_assignment, desc(assingment_prob)) %>% 
  ungroup() %>% 
  mutate(id = forcats::fct_inorder(factor(id)))

# Make a plot
p <- ggplot(plot_data, aes(id, prob, fill = pop)) +
  geom_col() +
  theme(strip.text.y = element_text(angle=90),
        axis.title.x = element_blank(),
        axis.text.x = element_blank(),
        strip.text.x = element_text(angle=90),
        axis.ticks.x = element_blank()) +
  facet_grid(~Population, scales = 'free', space = 'free')

p

```

Type `ggsave(plot = p, width = 30, height = 4,filename = "admixture_plot_HGDP_chr6_ordered_ancestry.pdf")` to save the plot as a pdf.

Look at patterns across populations. Do they follow a geographic structure? Is there a sign of Admixture?

Plotting admixture results with R might be tricky when many K's and Runs are present, but we can use [PONG](https://github.com/ramachandran-lab/pong) to visualize our ADMIXTURE runs and to summarize them. This program will take care of the colors and order of individuals and populations for us. It's great! If you ever have to use ADMIXTURE I highly recommend it.


*************************************************************

<br /><br /><br />

# Reference

Majority of the text and many figures were taken from 

Jobling, Mark, et al. "HUMAN EVOLUTIONARY GENETICS Second Edition." HUMAN EVOLUTIONARY GENETICS, SECOND EDITION (2014): 601-608.

And various other sources:
https://docs.readthedocs.io/en/stable/gsoc.html
https://comppopgenworkshop2019.readthedocs.io/en/latest/contents/03_pca/pca.html
https://www.cog-genomics.org/plink/1.9
https://www.cyberciti.biz/faq/how-to-tar-a-file-in-linux-using-command-line/
http://zzz.bwh.harvard.edu/plink/data.shtml#ped
https://www.cog-genomics.org/plink/1.9/formats#map
https://www.mv.helsinki.fi/home/mjxpirin/GWAS_course/material/GWAS5.html
https://www.sciencedirect.com/science/article/pii/S0002929714004273 
https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4001853/
https://github.com/GELOG/adam-ibs/wiki/Plink-IBS-MDS-Tutorial 

Marees, Andries T., et al. "A tutorial on conducting genome-wide association studies: Quality control and statistical analysis." International journal of methods in psychiatric research 27.2 (2018): e1608.